pkg/fanal/image/remote.go

```diff
@@ -67,16 +67,16 @@ func tryRemote(ctx context.Context, imageName string, ref name.Reference, option
 
 }
 
-func parsePlatform(ref name.Reference, p string) (*v1.Platform, error) {
+func parsePlatform(ref name.Reference, options []remote.Option, p string) (*v1.Platform, error) {
 	// OS wildcard, implicitly pick up the first os found in the image list.
 	// e.g. */amd64
 	if strings.HasPrefix(p, "*/") {
-		index, err := remote.Index(ref)
+		index, err := remote.Index(ref, options...)
 		if err != nil {
 			// Not a multi-arch image
 			if _, ok := err.(*remote.ErrSchema1); ok {
 				log.Logger.Debug("Ignored --platform as the image is not multi-arch")
-				return nil, nil
+				return nil, err
 			}
```

Comment:
It is my mistake. We don't need to handle ErrSchema1 here. I would rather have handled the following cases.
https://github.com/google/go-containerregistry/blob/6ba20c80c9580de62b0793d201dff23b22ded239/pkg/v1/remote/descriptor.go#L181-L183

But do you want to return an error anyway if an image is not multi-arch though `--platform` is specified?

Seems like Docker just ignores `--platform` if it is not multi-arch.

```
$ docker pull --platform amd64/linux amd64/alpine:3.16
3.16: Pulling from amd64/alpine
Digest: sha256:3d426b0bfc361d6e8303f51459f17782b219dece42a1c7fe463b6014b189c86d
Status: Image is up to date for amd64/alpine:3.16
docker.io/amd64/alpine:3.16
$ docker pull --platform arm64/linux amd64/alpine:3.16
3.16: Pulling from amd64/alpine
Digest: sha256:3d426b0bfc361d6e8303f51459f17782b219dece42a1c7fe463b6014b189c86d
Status: Image is up to date for amd64/alpine:3.16
docker.io/amd64/alpine:3.16
$ docker pull --platform arm64/windows amd64/alpine:3.16
3.16: Pulling from amd64/alpine
Digest: sha256:3d426b0bfc361d6e8303f51459f17782b219dece42a1c7fe463b6014b189c86d
Status: Image is up to date for amd64/alpine:3.16
docker.io/amd64/alpine:3.16
```

Reply:
Hi @knqyf263 

We are using this error for non multi-arch to fall back to aqua scanner. So, we can leave this error.

Reply:
As I said, we don't have to handle it. We can remove the following if-statement and just return an error.

```
if _, ok := err.(*remote.ErrSchema1); ok {
```

Reply:
@knqyf263 above ErrSchema1 check is removed and just returning the error. Updated the PR

Reply:
@knqyf263, One thing we have noted distinctively is that in the parsePlatform error throwing due to Schema1 manifest is resulting in the scan to completely discontinue. Is this a valid assumption that just with Schema1 manifest the image is not multi-arch image?

We have many images which our Aqua private registries failing with this error because of Schema1 manifest but they are falling back to Aqua scanner to successfully scan the actual multi-arch image.

The point is that the image(s) are multi-arch available in the registry but Schema1 manifest might not give us the needed information for image architecture details available under the registry.

For example,

Docker v2 api(HTTP)

"/v2/<repo_name>/manifests/<image_digest>"

gives us the image manifest in many cases and upon parsing the results we obtain the platforms(image-architectures)

![Screenshot from 2023-01-03 13-41-03](https://user-images.githubusercontent.com/45425247/210320141-739c99ed-6d8a-4f8c-817f-1221c1057387.png)

Reply:
>Is this a valid assumption that just with Schema1 manifest the image is not multi-arch image?

AFAIK, manifests for multi-arch images must be `application/vnd.oci.image.index.v1+json` or `application/vnd.docker.distribution.manifest.list.v2+json`. 
https://github.com/google/go-containerregistry/blob/6ba20c80c9580de62b0793d201dff23b22ded239/pkg/v1/remote/descriptor.go#L151-L153

The schema v1 here means `application/vnd.docker.distribution.manifest.v1+json` or `application/vnd.docker.distribution.manifest.v1+prettyjws`, but they don't support multi-arch. Please correct me if I'm wrong.
https://github.com/google/go-containerregistry/blob/6ba20c80c9580de62b0793d201dff23b22ded239/pkg/v1/remote/descriptor.go#L147-L150

I think we can assume images with schema v1 are not multi-arch.

Reply:
**We can assume schema v1 images are non multi-arch.**

We, in Aqua scanning flow try to get the image manifest by making the request to "/v2/<repo_name>/manifests/<image_digest>" with acceptable mediaTypes

**"application/vnd.docker.distribution.manifest.list.v2+json"
"application/vnd.docker.distribution.manifest.v2+json"
"application/vnd.docker.distribution.manifest.v1+json"
"application/vnd.oci.image.manifest.v1+json"**

reason being, YES we may not get the list of available platforms for a given image if we don't receive manifest response of "application/vnd.docker.distribution.manifest.list.v2+json"

but, "application/vnd.docker.distribution.manifest.v2+json" - constructs to schema2.Manifest(github.com/docker/distribution)
"application/vnd.docker.distribution.manifest.v1+json" - constructs to schema1.Manifest(github.com/docker/distribution)

These manifest types require additional call to
**"/v2/<repo_name>/blobs/<image_digest>"** in addition to **"/v2/<repo_name>/manifests/<image_digest>"** made above.

And **"/v2/<repo_name>/blobs/<image_digest>"** gives image config which contains the OS and Image architecture information with which we can proceed further.
 
but looks like the idea of determining the OS from parsePlatform which uses go-container registry only accepts **"application/vnd.docker.distribution.manifest.list.v2+json"
**"application/vnd.oci.image.manifest.v1+json"**
and determines, OS information cannot be obtained and fails the scan.

Reply:
First of all, determining OS from the config is completely different from the current approach. It currently uses os/arch in the manifest list.
If an image is not multi-arch, we can just ignore `--platform`. Do you want to fail scanning if the specified OS is different from OS in the config? I'm not sure about your expectation.

go-containerregistry doesn't support `application/vnd.docker.distribution.manifest.v1+json` and `application/vnd.docker.distribution.manifest.v1+prettyjws`. It just fails to scan images with the schema v1 (non multi-arch and non v2 schema).

Reply:
You requested my review, but I still don't get your point. How do you want to improve it?